google/cloud/storage/examples/storage_async_samples.cc

```diff
@@ -329,6 +329,81 @@ void StartBufferedUpload(
+         std::string object_name,
+         std::string destination_name) -> g::future<gcs::ObjectMetadata> {
+    auto [rewriter, token] = client.StartRewrite(bucket_name, object_name,
+                                                 bucket_name, destination_name);
+    while (token.valid()) {
+      auto [progress, t] =
+          (co_await rewriter.Iterate(std::move(token))).value();
+      token = std::move(t);
+      std::cout << progress.total_bytes_rewritten << " of "
+                << progress.object_size << " bytes rewritten\n";
+      if (progress.metadata) co_return *std::move(progress.metadata);
+    }
+    throw std::runtime_error("rewrite failed before completion");
+  };
+  //! [rewrite-object]
+
+  // The example is easier to test and run if we call the coroutine and block
+  // until it completes.
+  auto const metadata = coro(client, argv.at(0), argv.at(1), argv.at(2)).get();
+  std::cout << "Object successfully rewritten " << metadata << "\n";
+}
+
+void ResumeRewrite(google::cloud::storage_experimental::AsyncClient& client,
+                   std::vector<std::string> const& argv) {
+  //! [resume-rewrite]
+  namespace g = google::cloud;
+  namespace gcs = g::storage;
+  namespace gcs_ex = g::storage_experimental;
+  auto start = [](gcs_ex::AsyncClient& client, std::string bucket_name,
+                  std::string object_name,
+                  std::string destination_name) -> g::future<std::string> {
+    // First start a rewrite. In this example we will limit the number of bytes
+    // rewritten by each iteration to capture and resume the
+    auto [rewriter, token] = client.StartRewrite(
+        bucket_name, object_name, bucket_name, destination_name,
+        gcs::MaxBytesRewrittenPerCall(1024 * 1024));
+    auto [progress, t] = (co_await rewriter.Iterate(std::move(token))).value();
+    co_return progress.rewrite_token;
+  };
+  auto resume =
+      [](gcs_ex::AsyncClient& client, std::string bucket_name,
+         std::string object_name, std::string destination_name,
+         std::string rewrite_token) -> g::future<gcs::ObjectMetadata> {
+    // Continue rewriting, this could happen on a separate process, or even
+    // after the application restarts.
+    auto [rewriter, token] = client.ResumeRewrite(
+        bucket_name, object_name, bucket_name, destination_name, rewrite_token,
+        gcs::MaxBytesRewrittenPerCall(1024 * 1024));
+    while (token.valid()) {
+      auto [progress, t] =
+          (co_await rewriter.Iterate(std::move(token))).value();
+      token = std::move(t);
+      std::cout << progress.total_bytes_rewritten << " of "
+                << progress.object_size << " bytes rewritten\n";
+      if (progress.metadata) co_return *std::move(progress.metadata);
+    }
+  };
+  //! [resume-rewrite]
+
+  // The example is easier to test and run if we call the coroutine and block
+  // until it completes.
+  auto const rt = start(client, argv.at(0), argv.at(1), argv.at(2)).get();
```

Comment:
Just curious, what happens if `start(...)` is a complete call? Will `resume(...)` redo the copy?

(I see that we make sure `start(...)` is incomplete in our CI. Just wondering what a customer might run into if they don't do the same thing. Also, I will go away if you say: "who cares, this is an example")

Reply:
I *think* (but I have not tested this) the service will keep tokens for up to 1 week, and return the result of copying the object until the token expires.  It returns an error (410 Gone in HTTP, I assume NOT_FOUND in gRPC) if the token has been garbage collected.

Resuming with an empty token?  It will probably just rewrite again.